### PR TITLE
Support specifying Tenant's Domains

### PR DESCRIPTION
**Pull request by dvaldivia**
Introduces the field `.spec.features.domains` on the tenant which has an array for a list of domains for `minio` and `console`.

This helps configure the `MINIO_DOMAIN` environment variable properly, and works in conjunction with `.spec.features.bucketDNS` so both things can be properly enabled.

To test, apply the CRD on this PR and deploy the container image, then on a new or existing tenant add:

```
  features:
    domains:
      minio:
        - ns-1.cloud.lulu
```
or mix with bucketDNS
```
  features:
    bucketDNS: true
    domains:
      minio:
        - ns-1.cloud.lulu
```

you will be able to observe the `env` for `MINIO_DOMAIN` with the proper value

Signed-off-by: Daniel Valdivia <18384552+dvaldivia@users.noreply.github.com>

### CODE REVIEW

**Comment by harshavardhana** [comment]
@dvaldivia is this PR needed ?

**Comment by dvaldivia** [comment]
@harshavardhana yes! still a pain point for those who configure bucket dns internally and externally, this PR solves that

File: pkg/resources/statefulsets/minio-statefulset.go [hunk 1]
```diff
@@ -121,6 +120,17 @@ func minioEnvironmentVars(t *miniov2.Tenant, wsSecret *v1.Secret, hostsTemplate
 			},
 		})
 	}
+	// Check if any domains are configured
+	if t.Spec.Features != nil && t.Spec.Features.Domains != nil && len(t.Spec.Features.Domains.Minio) > 0 {
```
Comment (Alevsk) [inline pkg/resources/statefulsets/minio-statefulset.go:124]: we can abstract this to a `HasDomains` function or something similar

**Review by Alevsk — APPROVED** [review]

File: pkg/apis/minio.min.io/v2/types.go [hunk 2]
```diff
@@ -196,10 +221,16 @@ type TenantSpec struct {
...
+	// *Deprecated in Operator v4.3.2* +
 	//
 	// S3 related features can be disabled or enabled such as `bucketDNS` etc.
 	S3 *S3Features `json:"s3,omitempty"`
 	// *Optional* +
+	// *Deprecated in Operator v4.3.2* +
+	//
+	// S3 related features can be disabled or enabled such as `bucketDNS` etc.
+	Features *Features `json:"features,omitempty"`
+	// *Optional* +
```
Comment (harshavardhana) [inline pkg/apis/minio.min.io/v2/types.go:233]: this is not deprecated right? this is a new feature

File: pkg/apis/minio.min.io/v2/types.go [hunk 3]
```diff
@@ -56,15 +56,40 @@ type TenantScheduler struct {
...
 
-// S3Features (`s3`) - Object describing which S3 features to enable/disable in the MinIO Tenant. +
+// TenantDomains (`domains`) - List of domains used to access the tenant from outside the kubernetes clusters.
+// this will only configure MinIO for the domains listed, but external DNS configuration is still needed.
+type TenantDomains struct {
+	// List of Domains used by MinIO. This will enable DNS style access to the object store where the bucket name is
+	// inferred from a subdomain in the domain.
+	Minio []string `json:"minio,omitempty"`
+	// List of Domains used by Console
+	Console []string `json:"console,omitempty"`
```
Comment (harshavardhana) [inline pkg/apis/minio.min.io/v2/types.go:66]: don not who is using Console array here, can we only add it when needed?

File: pkg/apis/minio.min.io/v2/types.go [hunk 4]
```diff
@@ -56,15 +56,40 @@ type TenantScheduler struct {
...
-	BucketDNS bool `json:"bucketDNS"`
+	BucketDNS bool `json:"bucketDNS,omitempty"`
+}
+
+// Features (`features`) - Object describing which MinIO features to enable/disable in the MinIO Tenant. +
+//
+type Features struct {
+	// *Optional* +
+	//
+	//Specify `true` to allow clients to access buckets using the DNS path `<bucket>.minio.default.svc.cluster.local`. Defaults to `false`.
```
Comment (harshavardhana) [inline pkg/apis/minio.min.io/v2/types.go:85]:
```suggestion
	// Specify `true` to allow clients to access buckets using the DNS path `<bucket>.minio.default.svc.cluster.local`. Defaults to `false`.
```

File: Makefile [hunk 5]
```diff
@@ -73,17 +75,17 @@ plugin: regen-crd
 	@(cd $(PLUGIN_HOME); \
 		go vet ./... && \
 		go test -race ./... && \
-		GO111MODULE=on ${GOPATH}/bin/golangci-lint cache clean && \
-		GO111MODULE=on ${GOPATH}/bin/golangci-lint run --timeout=5m --config ../.golangci.yml)
+		GO111MODULE=on golangci-lint cache clean && \
+		GO111MODULE=on golangci-lint run --timeout=5m --config ../.golangci.yml)
```
Comment (harshavardhana) [inline Makefile:79]: this assumes `golangci-lint` is in PATH - this may not be the case since GOPATH might not be in PATH

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: Makefile (hunk 5 above)
Comment (dvaldivia) [reply Makefile:79]: in this makefile we have it like this in other places, like in `getdeps`, I think assuming it's in the PATH should be fine, as long as $GOPATH is in the PATH hahaha

File: pkg/apis/minio.min.io/v2/types.go (hunk 3 above)
Comment (dvaldivia) [reply pkg/apis/minio.min.io/v2/types.go:66]: the next PR will use it, but I'll add it on the next PR

File: pkg/apis/minio.min.io/v2/types.go (hunk 4 above)
Comment (dvaldivia) [reply pkg/apis/minio.min.io/v2/types.go:85]: done

File: pkg/apis/minio.min.io/v2/types.go (hunk 2 above)
Comment (dvaldivia) [reply pkg/apis/minio.min.io/v2/types.go:233]: nice catch

File: pkg/resources/statefulsets/minio-statefulset.go (hunk 1 above)
Comment (dvaldivia) [reply pkg/resources/statefulsets/minio-statefulset.go:124]: done

**Review by Alevsk — APPROVED** [review]

**Comment by harshavardhana** [comment]
Some CRD issues @dvaldivia 
```
Checking dependencies
go: downloading github.com/stretchr/testify v1.7.0
go: downloading github.com/pmezard/go-difflib v1.0.0
# github.com/minio/operator/pkg/apis/minio.min.io/v2
pkg/apis/minio.min.io/v2/zz_generated.deepcopy.go:694:7: in.Console undefined (type *TenantDomains has no field or method Console)
pkg/apis/minio.min.io/v2/zz_generated.deepcopy.go:695:17: in.Console undefined (type *TenantDomains has no field or method Console)
pkg/apis/minio.min.io/v2/zz_generated.deepcopy.go:695:31: out.Console undefined (type *TenantDomains has no field or method Console)
# github.com/minio/operator/pkg/apis/minio.min.io/v2
vet: pkg/apis/minio.min.io/v2/zz_generated.deepcopy.go:694:8: in.Console undefined (type *TenantDomains has no field or method Console)
make: *** [Makefile:48: govet] Error 2
Error: Process completed with exit code 2.```

**Comment by harshavardhana** [comment]
Comment missing @dvaldivia 
```
Running lint check
pkg/apis/minio.min.io/v2/helper.go:941:1: exported: comment on exported method Tenant.HasMinIODomains should be of the form "HasMinIODomains ..." (revive)
// HasDomains indicates whether domains are being specified for MinIO
^
make: *** [Makefile:45: lint] Error 1
```

**Comment by harshavardhana** [comment]
@dvaldivia can you rebase this PR when you get time?

**Comment by dvaldivia** [comment]
@harshavardhana done rebasing

**Review by harshavardhana — APPROVED** [review]

**Comment by harshavardhana** [comment]
PTAL @bexsoft @Alevsk @kanagarajkm

**Review by kanagarajkm — APPROVED** [review]

**Review by sinhaashish — APPROVED** [review]

**Pull request merged**